include/rmm/mr/device/limiting_resource_adaptor.hpp

```diff
@@ -0,0 +1,210 @@
+   * @param allocation_limit Maximum memory allowed for this allocator.
+   */
+  limiting_resource_adaptor(Upstream* upstream,
+                            std::size_t allocation_limit_ = std::numeric_limits<std::size_t>::max(),
+                            std::size_t allocation_alignment_ = 256)
+    : upstream_{upstream},
+      allocation_limit{allocation_limit_},
+      allocation_alignment(allocation_alignment_),
+      used(0)
+  {
+    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
+  }
+
+  limiting_resource_adaptor()                                 = delete;
+  ~limiting_resource_adaptor()                                = default;
+  limiting_resource_adaptor(limiting_resource_adaptor const&) = delete;
+  limiting_resource_adaptor(limiting_resource_adaptor&&)      = default;
+  limiting_resource_adaptor& operator=(limiting_resource_adaptor const&) = delete;
+  limiting_resource_adaptor& operator=(limiting_resource_adaptor&&) = default;
+
+  /**
+   * @brief Return pointer to the upstream resource.
+   *
+   * @return Upstream* Pointer to the upstream resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @brief Checks whether the upstream resource supports streams.
+   *
+   * @return true The upstream resource supports streams
+   * @return false The upstream resource does not support streams.
+   */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+  std::size_t space_free() const { return allocation_limit - used; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = nullptr;
+
+    std::size_t proposed_size = rmm::detail::align_up(bytes, allocation_alignment);
+    if (proposed_size + used <= allocation_limit) {
+      used += proposed_size;
+      try {
+        p = upstream_->allocate(bytes, stream);
+      } catch (std::exception const& e) {
+        RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE {}]",
+                      reinterpret_cast<void*>(stream),
+                      bytes,
+                      e.what());
+        throw;
```

Comment:
I don't think it should be the responsibility of this adaptor to log exceptions thrown from the upstream.